Source/WebKit/UIProcess/API/glib/WebKitSettings.cpp

```diff
@@ -4131,3 +4131,81 @@ WebKitFeatureList* webkit_settings_get_development_features(void)
 {
     return webkitFeatureListCreate(WebPreferences::internalDebugFeatures());
 }
+
+/**
+ * webkit_settings_apply_from_key_file:
+ * @settings: a #WebKitSettings
+ * @key_file: a #GKeyFile
+ * @group_name: Name of the group to read from @key_file
+ * @error: return location for error or %NULL to ignore
+ *
+ * Reads the contents of the given @group_name from the given @key_file and apply the value of
+ * each key/value to the corresponding property on the @settings.
+ *
+ * Returns: %TRUE if the settings were correctly applied or %FALSE otherwise.
+ *
+ * Since: 2.46
+ */
+gboolean webkit_settings_apply_from_key_file(WebKitSettings* settings, GKeyFile* keyFile, const gchar* groupName, GError ** error)
+{
+    g_return_val_if_fail(WEBKIT_IS_SETTINGS(settings), FALSE);
+    g_return_val_if_fail(keyFile, FALSE);
+
+    GObjectClass* klass = G_OBJECT_GET_CLASS(settings);
+    unsigned totalProperties = 0;
+    GUniquePtr<GParamSpec*> properties(g_object_class_list_properties(klass, &totalProperties));
+
+    if (!g_key_file_has_group(keyFile, groupName)) {
+        g_set_error(error, G_KEY_FILE_ERROR, G_KEY_FILE_ERROR_INVALID_VALUE, "The key file has no %s group", groupName);
+        return FALSE;
+    }
+
+    for (unsigned i = 0; i < totalProperties; i++) {
+        GUniqueOutPtr<GError> lookupError;
+        const char* propertyName = properties.get()[i]->name;
+        if (!g_key_file_has_key(keyFile, groupName, propertyName, &lookupError.outPtr())) {
+            if (lookupError) {
+                g_propagate_error(error, lookupError.release());
+                return FALSE;
+            }
+            continue; // Setting missing in GKeyFile, skip it.
```

Comment:
Setting missing in key file is not what makes `g_key_file_has_key` to fail?

Reply:
Missing keys cause the function to return `FALSE` and leave `error` untouched. As per [the docs](https://docs.gtk.org/glib/method.KeyFile.has_key.html):

> Note that this function does not follow the rules for GError strictly; the return value both carries meaning and signals an error. To use this function, you must pass a GError pointer in error, and check whether it is not NULL to see if an error occurred.

Meaning:

- Result is `TRUE` → no error, key exists.
- Result is `FALSE`:
  - If error is `NULL` → no error, key does not exist.
  - If error is non-`NULL` → error, we don't know if key exists or not.